### PR TITLE
Experimental Prerender API

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- This branch implements [the Prerender RFC](https://github.com/withastro/rfcs/pull/362), which allows Astro servers deployed using `output: 'server'` to prerender any page to static HTML.
- Introduces a new `experimental` flag, `prerender` to opt-in to this behavior.
- Page-level prerendering can be enabled by adding `export const prerender = true;` to any file in the `pages/` directory. These assets will be statically rendered at build time and included as assets.
- Reformats the build output to bring us closer to a one-file-per-route approach. Still more TODO, but this is a first step. 👀

## Testing

Tests have been added to verify this mode's output and the shape of the manifest. It's a bit difficult to test directly since our adapters are not responsible for serving static assets.

## Docs

Still a WIP, will be working on these after the change lands!

### CODE REVIEW

File: packages/astro/src/core/build/common.ts [hunk 1]
```diff
@@ -7,7 +8,11 @@ const STATUS_CODE_PAGES = new Set(['/404', '/500']);
 const FALLBACK_OUT_DIR_NAME = './.astro/';
 
 function getOutRoot(astroConfig: AstroConfig): URL {
-	return new URL('./', astroConfig.outDir);
+	if (astroConfig.output === 'static') {
+		return new URL('./', astroConfig.outDir);
+	} else {
+		return new URL('./', astroConfig.build.client);
```
Comment (matthewp) [inline packages/astro/src/core/build/common.ts:14]: isn't there a getClientRoot() in this file? this function was intended to return the actual root.

File: packages/astro/src/core/build/static-build.ts [hunk 2]
```diff
@@ -130,10 +136,27 @@ async function ssrBuild(opts: StaticBuildOptions, internals: BuildInternals, inp
 				input: [],
 				output: {
 					format: 'esm',
-					chunkFileNames: 'chunks/[name].[hash].mjs',
+					chunkFileNames: (chunkInfo) => {
+						if (chunkInfo.name.startsWith('pages/')) {
+							return `[name].[hash].mjs`;
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:141]: maybe in a pages/ subfolder so it's neater?

File: packages/astro/src/core/build/static-build.ts [hunk 3]
```diff
@@ -241,7 +264,57 @@ async function clientBuild(
...
+	// The SSR output is all .mjs files, the client output is not.
+	const files = await glob('**/*.mjs', {
+		cwd: fileURLToPath(out),
+	});
+	if (files.length) {
+		// Remove all the SSR generated .mjs files
+		await Promise.all(
+			files.map(async (filename) => {
+				const url = new URL(filename, out);
+				if (!allStaticFiles.has(filename.replace('server/', ''))) {
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:282]: fine for the draft, but before merged we shouldn't assume it's `server/`

File: packages/astro/src/core/build/static-build.ts [hunk 4]
```diff
@@ -241,7 +264,57 @@ async function clientBuild(
...
+					return;
+				}
+				await fs.promises.writeFile(url, 'export const _ = null;', { encoding: 'utf-8' });
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:285]: a comment on what this is for would be good

File: packages/astro/src/core/build/static-build.ts (hunk 2 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:141]: Ah sorry this is not very clear, but yes that's what this does.

File: packages/astro/src/core/build/common.ts (hunk 1 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/common.ts:14]: I don't see a `getClientRoot` function anywhere...

**Comment by natemoo-re** [comment]
!preview prerender

**Comment by natemoo-re** [comment]
!preview prerender

**Comment by natemoo-re** [comment]
!preview prerender

File: packages/astro/src/core/build/internal.ts [hunk 5]
```diff
@@ -189,6 +202,31 @@ export function* eachPageData(internals: BuildInternals) {
...
+}
+
+export function* eachServerPageData(internals: BuildInternals) {
+	for (const [id, pageData] of internals.pagesByViteID.entries()) {
+		if (!internals.pageOptionsByPage.get(id)?.prerender) {
+			yield pageData;
+		}
+	}
+}
+
```
Comment (natemoo-re) [inline packages/astro/src/core/build/internal.ts:229]: The Build API gets two new utilities which map over prerendered pages and server-only pages.

File: packages/astro/src/core/build/static-build.ts [hunk 6]
```diff
@@ -88,15 +93,33 @@ export async function staticBuild(opts: StaticBuildOptions) {
...
+		switch (settings.config.output) {
+			case 'static': {
+				await generatePages(opts, internals);
+				await cleanServerOutput(opts);
+				return;
+			}
+			case 'server': {
+				await injectManifest(opts, internals);
+				await generatePages(opts, internals);
+				await cleanStaticOutput(opts, internals);
```
Comment (natemoo-re) [inline packages/astro/src/core/build/static-build.ts:117]: The `server` output mode can now generate static pages as well!

File: packages/astro/src/core/build/static-build.ts [hunk 7]
```diff
@@ -125,6 +148,22 @@ async function ssrBuild(opts: StaticBuildOptions, internals: BuildInternals, inp
...
+						if (pageInfo) {
+							// prerendered pages should be split into their own chunk
+							// Important: this can't be in the `pages/` directory!
+							if (api.getModuleInfo(id)?.meta.astro?.pageOptions?.prerender) {
+								return `prerender`;
+							}
+							// pages should go in their own chunks/pages/* directory
+							return `pages${pageInfo.route.route.replace(/\/$/, '/index')}`;
+						}
+					} : undefined,
```
Comment (natemoo-re) [inline packages/astro/src/core/build/static-build.ts:166]: Fairly substantial change that's more future-facing: we're being more prescriptive about the build output. This is much closer to a one-route-per-file output that might be useful in the future.

File: packages/astro/src/core/build/static-build.ts [hunk 8]
```diff
@@ -232,18 +276,39 @@ async function clientBuild(
...
 				const url = new URL(filename, out);
-				await fs.promises.rm(url);
+				const text = await fs.promises.readFile(url, { encoding: 'utf8' });
+				const [, exports] = eslexer.parse(text);
+				// Replace exports (prerendered pages) with a noop
+				let value = 'const noop = (() => {});';
+				for (const e of exports) {
+					value += `\nexport const ${e.n} = noop;`;
+				}
+				await fs.promises.writeFile(url, value, { encoding: 'utf8' });
```
Comment (natemoo-re) [inline packages/astro/src/core/build/static-build.ts:311]: This checks the `prerender` chunk and wipes out everything except the exported pages. This allows us to only run a single SSR build, but eliminate any static-only dependencies from the server before deploy.

This is also why we're writing all static pages to a single chunk. Rollup is smart enough to isolate all of their deps in this chunk as well.

File: packages/astro/src/vite-plugin-scanner/index.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+					if (expr !== 'true') {
+						error(
+					logging,
+					'scanner',
+					`${colors.yellow(id)}
+File contains a \`prerender\` export, but its value cannot be statically analyzed! Expected \`true\` but got \`${expr}\`.`
+				);
+					}
+				}
+			}
```
Comment (natemoo-re) [inline packages/astro/src/vite-plugin-scanner/index.ts:69]: This scans for `export const prerender`

File: .changeset/funny-waves-worry.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+> **Note**
+> This API is not yet stable and is subject to possible breaking changes!
+
+- Deploy an Astro server without sacrificing the speed or cacheability of static HTML.
+- The Prerender API allows you to statically prerender specific `pages/` at build time.
+
+**Usage**
+
+- First, enable `experimental: { prerender: true }` in your `astro.config.mjs` file.
```
Comment (matthewp) [inline .changeset/funny-waves-worry.md:16]: you should also be able to enable this through a CLI flag, ala `--experimental-prerender`.

File: packages/astro/src/@types/astro.ts [hunk 11]
```diff
@@ -901,6 +901,36 @@ export interface AstroUserConfig {
 	 */
 	experimentalErrorOverlay?: boolean;
```
Comment (matthewp) [inline packages/astro/src/@types/astro.ts:902]: Shouldn't this be moved into the experimental object? Or is @bholmesdev doing that in his PR?

File: packages/astro/src/core/build/generate.ts [hunk 12]
```diff
@@ -70,17 +75,37 @@ export function chunkIsPage(
...
 export async function generatePages(opts: StaticBuildOptions, internals: BuildInternals) {
 	const timer = performance.now();
-	info(opts.logging, null, `\n${bgGreen(black(' generating static routes '))}`);
-
 	const ssr = opts.settings.config.output === 'server';
 	const serverEntry = opts.buildConfig.serverEntry;
 	const outFolder = ssr ? opts.buildConfig.server : getOutDirWithinCwd(opts.settings.config.outDir);
+
+	if (opts.settings.config.experimental.prerender) {
+		// HACK: force output to be treated as ESM
```
Comment (matthewp) [inline packages/astro/src/core/build/generate.ts:83]: why is this needed?

File: packages/astro/src/core/build/static-build.ts [hunk 13]
```diff
@@ -125,6 +148,22 @@ async function ssrBuild(opts: StaticBuildOptions, internals: BuildInternals, inp
 					assetFileNames: 'assets/[name].[hash][extname]',
 					...viteConfig.build?.rollupOptions?.output,
 					entryFileNames: opts.buildConfig.serverEntry,
+					manualChunks: opts.settings.config.experimental.prerender ? (id, api) => {
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:151]: I think this should be done in a plugin rather than in this file, to improve readability. Like in a new prerender plugin.

Comment (matthewp) [reply packages/astro/src/core/build/static-build.ts:151]: Also I think another plugin is setting manualChunks, is this overriding that? I think in that plugin we try and check to see if manualChunks already exists and if so it call it first.

Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:151]: Good call, didn't even consider that. Will update.

File: packages/astro/src/core/build/generate.ts (hunk 12 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/generate.ts:83]: Test failures when the fixture did not use `"type": "module"`, but I should confirm that this is still needed.

File: packages/astro/src/@types/astro.ts (hunk 11 above)
Comment (natemoo-re) [reply packages/astro/src/@types/astro.ts:902]: I can do it here!

File: packages/astro/src/core/build/static-build.ts [hunk 14]
```diff
@@ -232,18 +276,39 @@ async function clientBuild(
...
+		await eslexer.init;
 		// Remove all the SSR generated .mjs files
 		await Promise.all(
 			files.map(async (filename) => {
+				if (!allStaticFiles.has(filename)) {
+					return;
+				}
 				const url = new URL(filename, out);
-				await fs.promises.rm(url);
+				const text = await fs.promises.readFile(url, { encoding: 'utf8' });
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:304]: Can this be done in a plugin as well? Probably just have a `vite-plugin-prerender` plugin.

I think this can probably be done in the `generateBundles` hook and not need to read/write from the file after its been written.

File: packages/astro/src/core/build/static-build.ts [hunk 15]
```diff
@@ -266,6 +331,54 @@ async function cleanSsrOutput(opts: StaticBuildOptions) {
 			})
 		);
 	}
+	try {
+		fs.rmSync(new URL('./package.json', out));
+	} catch (e) {}
+
+	// TODO: fix this!
+	// // Clean out directly if the outDir is outside of root
+	// if (out.toString() !== opts.settings.config.outDir.toString()) {
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:340]: left in code?

File: packages/astro/src/vite-plugin-scanner/index.ts [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+			if (!includesExport(code)) return;
+
+			await eslexer.init;
+			const [_, exports] = eslexer.parse(code, id);
+			let pageOptions: PageOptions = {};
+			for (const e of exports) {
+				if (BOOLEAN_EXPORTS.has(e.n)) {
+					const name = e.n as keyof PageOptions;
+					pageOptions[name] = true;
+					let expr = code.slice(e.le).trim().replace(/\=/, '').trim().split(/[;\n]/)[0];
```
Comment (matthewp) [inline packages/astro/src/vite-plugin-scanner/index.ts:59]: gulp.

would be good to add a comment on what this is doing.

**Review by matthewp — CHANGES_REQUESTED** [review]
In addition to the comments left, the RFC mentions tests for the vite-plugin-scanner which I don't see here yet. Would be good to add those too.

https://github.com/withastro/rfcs/blob/hybrid-output/proposals/0000-prerender-api.md#testing-strategy

File: packages/astro/src/core/build/generate.ts (hunk 12 above)
Comment (matthewp) [reply packages/astro/src/core/build/generate.ts:83]: I think the way we make sure it is run as ESM is by adding the .mjs extension. Has that changed?

Comment (natemoo-re) [reply packages/astro/src/core/build/generate.ts:83]: Nope, I think this was a weird edge case as I was moving things around. Going to remove and check CI.

File: .changeset/funny-waves-worry.md (hunk 10 above)
Comment (natemoo-re) [reply .changeset/funny-waves-worry.md]: Fixed in [9e4a190](https://github.com/withastro/astro/pull/5297/commits/9e4a19076d9fa6e803d440f543c386e940116e57)

File: packages/astro/src/@types/astro.ts (hunk 11 above)
Comment (natemoo-re) [reply packages/astro/src/@types/astro.ts]: Fixed in [9e4a190](https://github.com/withastro/astro/pull/5297/commits/9e4a19076d9fa6e803d440f543c386e940116e57)

File: packages/astro/src/vite-plugin-scanner/index.ts (hunk 16 above)
Comment (natemoo-re) [reply packages/astro/src/vite-plugin-scanner/index.ts]: Refactored in [08e2d98](https://github.com/withastro/astro/pull/5297/commits/08e2d985ee47e39e1e0136e766f652f88f3bfa7c) and included a comment!

File: packages/astro/src/core/build/static-build.ts (hunk 15 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts]: Removed in [0c9ea57](https://github.com/withastro/astro/pull/5297/commits/0c9ea57a8618dd19493da74838c46a2b295c2900)

File: packages/astro/src/core/build/static-build.ts (hunk 14 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:304]: This can't be done in a plugin because it has to happen _after_ the SSR build runs. We are generating the `prerender` chunk, running it to statically generate the pages, then doing this step to remove the code here before the server is deployed.

This chunk will only contain the code pertaining to the `prerender` chunk, so it is safe to modify.

Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:307]: Added some comments in [7aa1914](https://github.com/withastro/astro/pull/5297/commits/7aa191458d4ae2841cb19cac3b5abae3fe17bd3e)

File: packages/astro/src/core/build/static-build.ts (hunk 13 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts]: Refactored into `vite-plugin-prerender` in [ecc44ab](https://github.com/withastro/astro/pull/5297/commits/ecc44ab042d17b640dda3f3e600d46f2812e203d)

**Comment by natemoo-re** [comment]
!preview prerender

File: packages/astro/test/units/vite-plugin-scanner/scan.test.js [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+		expect(result.prerender).to.equal(true);
+	});
+
+	it('recognizes constant boolean literal (false)', async () => {
+		const result = await scan(`export const prerender = false;`, '/src/components/index.astro');
+		expect(result.prerender).to.equal(false);
+	});
+
+	it('recognizes let boolean literal', async () => {
+		const result = await scan(`export let prerender = true;`, '/src/components/index.astro');
```
Comment (matthewp) [inline packages/astro/test/units/vite-plugin-scanner/scan.test.js:21]: I thought we were going to make this only work with `const`. Using let someone might think they could do:

```js
export let prerender = true;

if(someCondition) {
  prerender = false;
}
```

File: packages/astro/src/core/build/static-build.ts (hunk 14 above)
Comment (matthewp) [reply packages/astro/src/core/build/static-build.ts:293]: Are you saying that it has to happen after *generation*?

Comment (matthewp) [reply packages/astro/src/core/build/static-build.ts:293]: I think I understand if that's the case.

File: packages/astro/src/core/build/static-build.ts [hunk 18]
```diff
@@ -232,7 +262,77 @@ async function clientBuild(
...
+		allStaticFiles.add(internals.pageToBundleMap.get(pageData.moduleSpecifier));
+	}
+	const ssr = opts.settings.config.output === 'server';
+	const out = ssr ? opts.buildConfig.server : getOutDirWithinCwd(opts.settings.config.outDir);
+	// The SSR output is all .mjs files, the client output is not.
+	const files = await glob('**/*.mjs', {
+		cwd: fileURLToPath(out),
+	});
+
+	if (files.length) {
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:281]: it looks like this will currently run even if the user isn't using prerendering. Could we check if `allStaticFiles.size` as well to make sure this part only runs on prerendering projects?

File: packages/astro/test/units/vite-plugin-scanner/scan.test.js (hunk 17 above)
Comment (natemoo-re) [reply packages/astro/test/units/vite-plugin-scanner/scan.test.js:21]: Yes, we did agree to that! I avoided doing it here because there wasn't an easy way to surface what type of variable you were using, but I'm sure I can figure something out.

File: packages/astro/src/core/build/static-build.ts (hunk 14 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:293]: Yes, _after generation_ is the right way to phrase it.

File: packages/astro/src/core/build/static-build.ts (hunk 18 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:281]: I just double checked, `cleanStaticOutput` only ever runs on prerendering projects, so no need for extra logic here.

File: packages/astro/test/units/vite-plugin-scanner/scan.test.js (hunk 17 above)
Comment (natemoo-re) [reply packages/astro/test/units/vite-plugin-scanner/scan.test.js]: Updated in [d1546a2](https://github.com/withastro/astro/pull/5297/commits/d1546a229436c2642b4c59e1092bde1019036964)

File: packages/astro/test/units/vite-plugin-scanner/scan.test.js [hunk 19]
```diff
@@ -17,9 +17,14 @@ describe('astro scan', () => {
 		expect(result.prerender).to.equal(false);
 	});
 
-	it('recognizes let boolean literal', async () => {
-		const result = await scan(`export let prerender = true;`, '/src/components/index.astro');
-		expect(result.prerender).to.equal(true);
+	it('throws on let boolean literal', async () => {
```
Comment (matthewp) [inline packages/astro/test/units/vite-plugin-scanner/scan.test.js:20]: can you add a test for `var` as well?

File: packages/astro/src/vite-plugin-scanner/index.ts [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+export default function astroScannerPlugin({ settings, logging }: { settings: AstroSettings, logging: LogOptions }): VitePlugin {
+	function normalizeFilename(filename: string) {
+		if (filename.startsWith('/@fs')) {
+			filename = filename.slice('/@fs'.length);
+		} else if (filename.startsWith('/') && !ancestor(filename, settings.config.root.pathname)) {
+			filename = new URL('.' + filename, settings.config.root).pathname;
+		}
+		return filename;
+	}
```
Comment (bholmesdev) [inline packages/astro/src/vite-plugin-scanner/index.ts:18]: Curious if we can reuse `normalizeFilename` from `vite-plugin-utils` here?

**Review by bholmesdev — APPROVED** [review]
Nonblocking comment on code reuse. But the features this unlocks have `pre`rendered me speechless 😶

File: packages/astro/test/units/vite-plugin-scanner/scan.test.js (hunk 19 above)
Comment (natemoo-re) [reply packages/astro/test/units/vite-plugin-scanner/scan.test.js:20]: Added in [b17b6dd](https://github.com/withastro/astro/pull/5297/commits/b17b6ddd00d3628b2f1c0104e90169224973f609)

**Review by matthewp — APPROVED** [review]

File: packages/astro/src/@types/astro.ts [hunk 21]
```diff
@@ -895,11 +896,40 @@ export interface AstroUserConfig {
...
+			*/
+		errorOverlay?: boolean;
+		/**
+		 * @docs
+		 * @name experimental.prerender
+		 * @type {boolean}
+		 * @default `false`
+		 * @version 1.7.0
+		 * @description
+		 * Enable experimental support for Prerendered pages when generating a server.
```
Comment (delucis) [inline packages/astro/src/@types/astro.ts:920]: No need to capitalize here I don’t think.

```suggestion
		 * Enable experimental support for prerendered pages when generating a server.
```

File: packages/astro/src/@types/astro.ts [hunk 22]
```diff
@@ -895,11 +896,40 @@ export interface AstroUserConfig {
...
+		 *
+		 * To enable this feature, set `experimental.prerender` to `true` in your Astro config:
+		 *
+		 * ```js
+		 * {
+		 * 	experimental: {
+		 *		prerender: true,
+		 * 	},
+		 * }
```
Comment (delucis) [inline packages/astro/src/@types/astro.ts:929]: Think we probably need the closing backticks or we’ll hit a render error when pulling this into docs.
```suggestion
		 * }
		 * ```
```

File: packages/astro/src/core/errors/errors-data.ts [hunk 23]
```diff
@@ -393,6 +393,22 @@ See https://docs.astro.build/en/guides/server-side-rendering/ for more informati
 			`Could not render \`${componentName}\`. No matching import has been found for \`${componentName}\`.`,
 		hint: 'Please make sure the component is properly imported.',
 	},
+	/**
+	 * @docs
+	 * @description
+	 * A `prerender` export was detected, but the value was not statically analyzable. Values computed at runtime are not supported, so `export const prerender` only supports the exact values `true` or `false`.
```
Comment (delucis) [inline packages/astro/src/core/errors/errors-data.ts:399]: Not a huge change here but small wording tweak:

```suggestion
	 * A `prerender` export was detected, but the value was not statically analyzable. Values computed at runtime are not supported, so `export const prerender` can only be set to `true` or `false`. Variables are not supported.
```

**Review by delucis — COMMENTED** [review]
Left a couple wording tweaks on the types/errors, but otherwise LGTM 👍

**Review by delucis — APPROVED** [review]
